docs/proposals/compute-resource-metrics-api.md

```diff
@@ -0,0 +1,179 @@
+# Kubernetes compute resource metrics API
+
+## Goals
+
+Provide resource usage metrics on pods and nodes on the API server to be used
+by the scheduler to improve job placement, utilization, etc. and by end users
+to understand the resource utilization of their jobs.
+
+## Current state
+
+Right now, the Kubelet exports container metrics via an API endpoint. This
+information is not gathered nor served by the Kubernetes API server.
+
+## Use cases
+
+The first user will be kubectl. The resource usage data can be shown to the
+user via a periodically refreshing interface similar to `top` on Unix-like
+systems. This info could let users assign resource limits more efficiently.
+
+```
+$ kubectl top kubernetes-minion-abcd
+Pod: monitoring-heapster-abcde
+CPU:
+    mean: 0.12 cores
+    95th: 1.10 cores
+    max:  1.91 cores
+Memory:
+    mean: 300MB (10%)
+    95th: 330MB (11%)
+    max:  395MB (13%)
+```
+
+A second user will be the scheduler. To assign pods to nodes efficiently, the
+scheduler needs to know the current free resources on each node.
+
+## Proposed endpoints
+
+	/api/v1/namespaces/myns/podMetrics/mypod/[derived|raw]
```

Comment:
Actually, if we ever want this to be backed by another component, we
probably want metrics to be in their own group.

On Jul 28, 2015, at 5:00 PM, Derek Carr <notifications@github.com> wrote:

In docs/proposals/compute-resource-metrics-api.md
<https://github.com/GoogleCloudPlatform/kubernetes/pull/11951#discussion_r35700234>
:

> +CPU:
> +    mean: 0.12 cores
> +    95th: 1.10 cores
> +    max:  1.91 cores
> +Memory:
> +    mean: 300MB (10%)
> +    95th: 330MB (11%)
> +    max:  395MB (13%)
> +```
> +
> +A second user will be the scheduler. To assign pods to nodes efficiently, the
> +scheduler needs to know the current free resources on each node.
> +
> +## Proposed endpoints
> +
> +	/api/v1/namespaces/myns/podMetrics/mypod/[derived|raw]

can you weigh the pros and cons of modeling this as a subresource of pod
instead of a sibling?

—
Reply to this email directly or view it on GitHub
<https://github.com/GoogleCloudPlatform/kubernetes/pull/11951/files#r35700234>
.